refactor: Replace chain relayTransactions/submitMemoryPool by higher method

ariard opened the pull request:
Remove CWalletTx::AcceptToMemoryPool

Replace CWalletTx::RelayWalletTransaction by SubmitMemoryPoolAndRelay


Add a relay flag to broadcastTransaction because wasn't sure of ReacceptWalletTransactions semantic.

Obviously, working on implementing https://github.com/bitcoin/bitcoin/pull/14978#issuecomment-459373984 to add the new higher-method in Node interface, will add a commit, just need more thought to do it cleanly

ryanofsky commented on src/interfaces/chain.cpp:
```diff
@@ -306,10 +301,58 @@ class ChainImpl : public Chain
         return it && (*it)->GetCountWithDescendants() > 1;
     }
-    void relayTransaction(const uint256& txid) override
+    bool broadcastTransaction(const CTransactionRef& tx, const uint256& txid, CValidationState& state, const CAmount& absurdfee, bool relay) override
```
Thanks for consolidating the code from `relayTransaction` and `submitToMemoryPool` in `interfaces/chain.cpp` and from `BroadcastTransaction` in `node/transaction.cpp`.

I do think the right place for all this code to live is in `node/transaction.cpp`, though. `interfaces/chain.cpp` is really just meant to be a simple glue file with lines like:

```c++
bool broadcastTransaction(...) override { return BroadcastTransaction(...); }
```

Also, I didn't look closely yet, but it seems like this refactoring might be changing wallet behavior (this could explain travis test failures). It would probably be better if this PR avoided changing behavior to the extent possible. This might require adding some extra options to the existing `BroadcastTransaction` function in `node/transaction.cpp`, or adding a second broadcast function there that can hopefully share some code with the current one.

ariard replied:
Okay was thinking than ```node/transaction.cpp``` was temporary and long term goal was to aggregate it to ```interfaces/node.cpp```.
Didn't want to change wallet behavior, fixed it by adding options in the existing BroadcastTransaction (but still not sure of ReacceptWalletTransactions, should we broadcast txn there ?), functional tests should be good now

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/15713#discussion_r271524444

Thanks! This looks good and now net removes code (+63/-83).

> Okay was thinking than node/transaction.cpp was temporary

Yeah, that may have been the original thought. But I think it could be a reasonable long term location if more things are moved around it. In any case it's where the BroadcastTransaction function currently lives, so it's easier not to move in this PR, which is more about simplifying wallet code.

> and long term goal was to aggregate it to interfaces/node.cpp

Everything in interfaces/, including interfaces/node.cpp is just supposed to be very short glue code (the files would be huge otherwise). The interfaces are a way of letting wallet, node, and gui code run in different processes, and a way of preventing them from accessing each others internal state.

jnewbery commented:
Thanks for looking at this @ariard. My view is different from @ryanofsky's. I think rather than trying to maintain existing behaviour, we should use this interface tidyup to better define the relationship between the wallet and the node, ie to answer the question "What services should the node offer to the wallet?" Currently, the node exposes two methods to the wallet for sending a transaction:

1. `submitToMemoryPool(const CTransactionRef& tx, CAmount absurd_fee, CValidationState& state)`: Add transaction to memory pool if the transaction fee is below the amount specified by absurd_fee. Returns false if the transaction could not be added due to the fee or for another reason. This is only called in `CWalletTx::AcceptToMemoryPool()`
2. `relayTransaction(const uint256& txid)`: adds the txid to the INV queue for each peer. This is only called in `CWalletTx::RelayWalletTransaction()` (which also calls `CWalletTx::AcceptToMemoryPool()`)

Note that to relay an INV for the transaction to peers (2), the transaction must already be in the mempool. See:

https://github.com/bitcoin/bitcoin/blob/ba54342c9dd3f2e5cdeed9ac57f1924f0d885cc6/src/net_processing.cpp#L3804

`relayTransaction()` is only ever called if `InMempool() || AcceptToMemoryPool()` is true.

The wallet currently calls `CWalletTx::AcceptToMemoryPool()` in the following places:

1a. In `CWallet::ReacceptWalletTransactions()`, called after the wallet has started up to make sure that its transactions are added to the mempool at start.
2a. In `CWalletTx::RelayWalletTransaction()`, mentioned above.
3a. In `CWallet::CommitTransaction()`, when a transaction is to be broadcast to the network for the first time.

The wallet currently calls `CWalletTx::RelayWalletTransaction()` in the following places:

2a. In `CWallet::ResendWalletTransactions()`, called on a timer to rebroadcast unconfirmed transactions.
2b. In `CWallet::CommitTransaction()`, but only if the transaction is successfully added to the mempool.

I've been thinking about this in the context of the wallet rebroadcasting transactions (see discussion here: http://www.erisian.com.au/meetbot/bitcoin-core-dev/2019/bitcoin-core-dev.2019-03-29-19.10.log.html#l-9). My view now is that the wallet should not have any access to relay functionality. It should be able to submit transactions to the node's mempool, and it should then be the mempool's responsibility to broadcast/rebroadcast the transaction (see comment here: http://www.erisian.com.au/meetbot/bitcoin-core-dev/2019/bitcoin-core-dev.2019-03-29-19.10.log.html#l-92). That seems like a much cleaner divide of responsibility.

This PR seems like a simplification in terms of combining `submitToMemoryPool()` and `relayTransaction()` into a single method, but I don't think that it moves us towards a point where the node<->wallet interface is simpler to understand.

ryanofsky commented:
John are you just saying that this PR does not go far enough, or that it does something bad or makes things more complicated?

I haven't looked closely enough at this yet but if it exposes one node method node instead of two, and reduces the amount of code without changing behavior, it would seem like a strict improvement. Is your rebroadcast proposal harder to implement after this PR?

jnewbery commented:
> I haven't looked closely enough at this yet but if it exposes one node method node instead of two, and reduces the amount of code without changing behavior, it would seem like a strict improvement.

It does reduce the number of node methods exposed, but does so by adding a flag to one of them which toggles the behaviour. I'm not sure if that counts as a simplification, or it's just moving the complexity around. I'm also not sure if this maintains existing behaviour. For example, the `GetDepthInMainChain(locked_chain) == 0` in `RelayWalletTransaction()` seems to have gone and not been replaced by anything.

ariard commented on src/interfaces/chain.cpp:
```diff
@@ -210,8 +205,7 @@ class NotificationsHandlerImpl : public Handler, CValidationInterface
         // should be able to be removed entirely if `ResendWalletTransactions`
         // is replaced by a wallet timer as suggested in
         // https://github.com/bitcoin/bitcoin/issues/15619
```
Done

ariard commented:
I kinda agree with @jnewbery here, wallet shouldn't have access to relay, because if so it may add complexity of proper tx rebroadcasting tracking?
Have taken GetDepthInMainChain and IsCoinbase out, because seems it's cleaner to let the mempool do the checks, in AcceptToMemoryPoolWorker if I understand it well. And not multiply spurious wallet accesses to chain state.
Sorry scope of PR wasn't clear, would happily change it to underline the real issue I guess.

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -1867,43 +1867,44 @@ void CWallet::ReacceptWalletTransactions(interfaces::Chain::Lock& locked_chain)
-        const uint256& wtxid = item.first;
-        CWalletTx& wtx = item.second;
-        assert(wtx.GetHash() == wtxid);
+        auto locked_chain = chain().lock();
```
Yes on my own, should be corrected!

ariard commented:
Okay will keep working on it by breaking up the PR and justify changes. Had a look on #15728, IMO conflicts should be solved smoothly once yours get into it.
@jnewbery  I've taken notes of your proposed changes of preventing wallet access to relay functionality but as far as I understand it would need first implementation of relay responsibility  in mempool, something not submit yet ? Once that done, removing relay from this code path seems really straightforward changes.

Commit:
Add BroadcastTransaction utility usage in Chain interface

Access through a broadcastTransaction method.
Add a wait_callback flag to turn off race protection when wallet
already track its tx being in mempool

Standardise highfee, absurdfee variable name to max_tx_fee

We drop the P2P check in BroadcastTransaction as g_connman is only
called by RPCs and the wallet scheduler, both of which are initialized
after g_connman is assigned and stopped before g_connman is reset.

Commit:
Introduce CWalletTx::SubmitMemoryPoolAndRelay

Higher wallet-tx method combining RelayWalletTransactions and
AcceptToMemoryPool, using new Chain::broadcastTransaction

Commit:
Remove duplicate checks in SubmitMemoryPoolAndRelay

IsCoinBase check is already performed early by
AcceptToMemoryPoolWorker
GetDepthInMainChain check is already perfomed by
BroadcastTransaction

To avoid deadlock we MUST keep lock order in
ResendWalletTransactions and CommitTransaction,
even if we lock cs_main again further.
in BroadcastTransaction. Lock order will need
to be clean at once in a future refactoring

Commit:
Remove unused submitToMemoryPool and relayTransactions Chain interfaces

ariard commented:
Rebased, splitted the commits, added justifications. Note that I removed relay flag in BroadcastTransaction, if I get John opinion submitting tx to memory pool should imply relay if it's a valid one.

jnewbery commented on src/interfaces/chain.cpp:
```diff
@@ -301,10 +296,13 @@ class ChainImpl : public Chain
         return it && (*it)->GetCountWithDescendants() > 1;
     }
-    void relayTransaction(const uint256& txid) override
+    bool broadcastTransaction(const CTransactionRef& tx, uint256& txid, std::string& err_string, const CAmount& absurdfee, bool wait_callback) override
```
This interface function doesn't need to expose the `wait_callback` bool since it's only ever called with `wait_callback` set to `false`.

ariard replied:
Ah yes, thanks done

jnewbery commented on src/interfaces/chain.cpp:
```diff
@@ -301,10 +296,13 @@ class ChainImpl : public Chain
-        CInv inv(MSG_TX, txid);
-        g_connman->ForEachNode([&inv](CNode* node) { node->PushInventory(inv); });
+        const TransactionError err = BroadcastTransaction(tx, txid, err_string, absurdfee, wait_callback);
+        if (TransactionError::OK == err) {
```
nit: `return TransactionError::OK == err;`

ariard replied:
Yes

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -1892,30 +1891,27 @@ void CWallet::ReacceptWalletTransactions(interfaces::Chain::Lock& locked_chain)
     // Don't relay abandoned transactions
     if (isAbandoned()) return false;
-    // Don't relay conflicted or already confirmed transactions
-    if (GetDepthInMainChain(locked_chain) != 0) return false;
```
There I still have a doubt on BroadcastTransaction HaveChain check behavior. ```HaveChain = !existingCoin.IsSpent()``` doesn't mean if tx is in chain, but all of its outputs has been spent, we accept it again to mempool/relay ?

Sjors replied:
`IsSpent()` is "misleadingly" named. Because we use `AccessCoin`: `Return a reference to Coin in the cache, or a pruned one if not found.`. Pruned in this case means all its `CTxOut` are `IsNull()`, and `IsSpent()` just calls `IsNull()`.

jnewbery replied:
Thanks @Sjors . I agree with you that `IsSpent()` is a misleading name. A more accurate name would be `DoesNotExist()`. Negating the function here determines whether the coin exists in the current UTXO set.

The check in `BroadcastTransaction()` is an optimization: if any of the outputs from the tx still exist in the UTXO set then the tx must be confirmed. If the coin has been confirmed but it's outputs don't exist, then it'll fail later (because its inputs can't exist if it has already been spent).

This check in `RelayWalletTransaction()` is also an optimization. If the transaction has been confirmed (or a conflicting transaction has been confirmed), then the block hash will be recorded in the `CMerkleTx`, and we can check in the chain whether that block exists.

An alternative to removing this call to `GetDepthInMainChain()` would be to merge https://github.com/bitcoin/bitcoin/pull/15931 first, which removes `GetDepthInMainChain`'s requirement on locking the chain.

jonatack replied:
> Thanks @Sjors . I agree with you that `IsSpent()` is a misleading name. A more accurate name would be `DoesNotExist()`. Negating the function here determines whether the coin exists in the current UTXO set.
> 
> The check in `BroadcastTransaction()` is an optimization: if any of the outputs from the tx still exist in the UTXO set then the tx must be confirmed. If the coin has been confirmed but it's outputs don't exist, then it'll fail later (because its inputs can't exist if it has already been spent).

It might be useful to enhance the code documentation added in 976d1e485 with this information (if it isn't expected to change soon). I found it v helpful to understanding the `!existingCoin.IsSpent())` optimisation.

jnewbery commented on src/node/transaction.cpp:
```diff
@@ -55,15 +55,17 @@ TransactionError BroadcastTransaction(const CTransactionRef tx, uint256& hashTx,
     } // cs_main
 
-    promise.get_future().wait();
+    if (wait_callback) {
```
I think you can make this slightly clearer by adding a local variable `callback_set{false}`, setting it to `true` in the first `else if (wait_callback)` block, removing the second `else if (wait_callback)` block, and then testing for it here.

ariard replied:
Okay, also add params comments on it in.h

jnewbery commented on src/interfaces/chain.h:
```diff
@@ -189,8 +180,10 @@ class Chain
+    //! Transaction is added to memory pool, if the transdaction fee is below the
+    //! amount specified by absurdfee, and broadcast to all nodes.
+    //! Return false if the transaction could not be added due to the fee or for another reason.
+    virtual bool broadcastTransaction(const CTransactionRef& tx, uint256& txid, std::string& err_string, const CAmount& absurdfee, bool wait_callback) = 0;
```
nit: I'd prefer to standardise on the name `max_tx_fee` (rather than `max_tx_fee`, `highfee`, `absurdfee`, etc)

jnewbery reviewed (commented):
> opinion submitting tx to memory pool should imply relay if it's a valid one

No, this would cause the wallet to rebroadcast all of its unconfirmed transactions at startup. That's a change in behaviour and a potentially bad privacy leak.

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -1892,30 +1891,27 @@ void CWallet::ReacceptWalletTransactions(interfaces::Chain::Lock& locked_chain)
+    // user could call sendmoney in a loop and hit spurious out of funds errors
+    // because we think that this newly generated transaction's change is
+    // unavailable as we're not yet aware that it is in the mempool.
+    if (!(fInMempool |= pwallet->chain().broadcastTransaction(tx, txid, err_string, pwallet->chain().maxTxFee(), false))) return false;
```
This is also a behaviour change because previously, `AcceptToMemoryPool()` would return `false` if `submitToMemoryPool()` returned `false`. Now, if `fInMempool` is set to `true` but `broadcastTransaction()` returns `false`, then it'll return `true`.

I think it makes sense to stick with the structure of the old `AcceptToMemoryPool()` function.

ariard replied:
Okay back to AcceptToMemoryPool structure

jnewbery commented on src/node/transaction.h:
```diff
@@ -16,9 +16,11 @@
- * @param[in]  highfee Reject txs with fees higher than this (if 0, accept any fee)
+ * @param[in]  max_tx_fee reject txs with fees higher than this (if 0, accept any fee)
+ * @param[in]  relay flag if both mempool insertion and p2p relay  are requested
+ * @param[in]  wait_callback, wait until callbacks have been processed to avoid stale result due to a sequentially RPC
```
Thanks for adding comments. I'd add to the description of the `wait_callback` param that this **must not** be set while cs_main, cs_mempool or cs_wallet is being held or there'll be a deadlock.

jnewbery reviewed (commented):
A few comments inline. The intermediate commit 07d5787041461bcfb8b5777809d874869710aa1e does not build because `locked_chain` is not defined in `SubmitMemoryPoolAndRelay()`.

ariard commented on src/interfaces/chain.cpp:
```diff
@@ -301,10 +296,10 @@ class ChainImpl : public Chain
-        CInv inv(MSG_TX, txid);
-        g_connman->ForEachNode([&inv](CNode* node) { node->PushInventory(inv); });
+        const TransactionError err = BroadcastTransaction(tx, txid, err_string, max_tx_fee, relay, false);
+	return TransactionError::OK == err;
```
Thanks, configured my editor accordingly!

ariard commented on src/node/transaction.h:
```diff
@@ -16,9 +16,11 @@
  * @param[out] &err_string reference to std::string to fill with error string if available
- * @param[in]  highfee Reject txs with fees higher than this (if 0, accept any fee)
+ * @param[in]  max_tx_fee reject txs with fees higher than this (if 0, accept any fee)
+ * @param[in]  relay flag if both mempool insertion and p2p relay  are requested
```
Corrected but was to align with other descriptions?

jnewbery replied:
sorry, I meant the two spaces between `relay` and `are`. I think two spaces at the front to align with the other params is fine.

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -1892,30 +1891,27 @@ void CWallet::ReacceptWalletTransactions(interfaces::Chain::Lock& locked_chain)
+    // user could call sendmoney in a loop and hit spurious out of funds errors
+    // because we think that this newly generated transaction's change is
+    // unavailable as we're not yet aware that it is in the mempool.
+    if (!(fInMempool |= pwallet->chain().broadcastTransaction(tx, txid, err_string, pwallet->chain().maxTxFee(), false))) return false;
```
What's your take on returning a pair of boolean from chain().broadcastTransaction() to encode both mempool/relay results ?

jnewbery replied:
In the existing code there's not a return code for `chain::relayTransaction()`. I think it's fine for the new `broadcastTransaction` interface method to only return `false` if the transaction failed to get into the mempool, and treat `TransactionError::P2P_DISABLED` as success. That way, `fInMempool` will get updated in exactly the same way as the existing code.

ariard replied:
Good, added ```TransactionError::P2P_DISABLED == err``` check in broadcastTransaction interface

ariard commented:
Thanks @jnewbery for the review and sorry for the ugly commit split. I think I've addressed all your points except this one : https://github.com/bitcoin/bitcoin/pull/15713#discussion_r274955618 ?

practicalswift commented on src/interfaces/chain.h:
```diff
@@ -189,8 +180,10 @@ class Chain
 
-    //! Relay transaction.
-    virtual void relayTransaction(const uint256& txid) = 0;
+    //! Transaction is added to memory pool, if the transdaction fee is below the
```
Nit: Should be "transaction" :-)

ariard replied:
Thanks, good catch!

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -1892,32 +1891,29 @@ void CWallet::ReacceptWalletTransactions(interfaces::Chain::Lock& locked_chain)
 }
 
-bool CWalletTx::RelayWalletTransaction(interfaces::Chain::Lock& locked_chain)
+bool CWalletTx::SubmitMemoryPoolAndRelay(std::string& err_string, bool relay, interfaces::Chain::Lock& locked_chain)
```
You now have `locked_chain` as a parameter, even though it's unused in this function in the final commit.

ariard replied:
Yes rebase issue, should be solved

jnewbery commented:
Still a couple of minor issues to fix up.

jnewbery commented on src/interfaces/chain.cpp:
```diff
@@ -30,6 +30,7 @@
 #include <util/system.h>
 #include <validation.h>
 #include <validationinterface.h>
+#include <node/transaction.h>
```
nit: sort includes

jnewbery commented on src/interfaces/chain.cpp:
```diff
@@ -301,10 +296,10 @@ class ChainImpl : public Chain
     {
-        CInv inv(MSG_TX, txid);
-        g_connman->ForEachNode([&inv](CNode* node) { node->PushInventory(inv); });
+        const TransactionError err = BroadcastTransaction(tx, txid, err_string, max_tx_fee, relay, false);
```
nit: could include an inline comment for the last parameter:

```c++
const TransactionError err = BroadcastTransaction(tx, txid, err_string, max_tx_fee, relay, /*wait_callback*/ false);
```

jnewbery commented on src/node/transaction.cpp:
```diff
@@ -52,27 +53,28 @@ TransactionError BroadcastTransaction(const CTransactionRef tx, uint256& hashTx,
+        promise.get_future().wait();
+    }
 
     if (!g_connman) {
```
Should this also be within the `if (relay)` block? Returning a `TransactionError::P2P_DISABLED` error when the caller hasn't asked for the transaction to be relayed seems wrong.

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -1892,32 +1891,29 @@ void CWallet::ReacceptWalletTransactions(interfaces::Chain::Lock& locked_chain)
-    return true;
+    // Submit transaction to mempool for relay
+    pwallet->WalletLogPrintf("Submitting wtx %s to mempool for relay\n", GetHash().ToString());
+    uint256 txid;
```
See comment in interface/chain.h. I recommend you remove the unused uint256 from the interface method.

jnewbery commented on src/node/transaction.cpp:
```diff
@@ -13,10 +13,11 @@
 #include <future>
 
-TransactionError BroadcastTransaction(const CTransactionRef tx, uint256& hashTx, std::string& err_string, const CAmount& highfee)
+TransactionError BroadcastTransaction(const CTransactionRef tx, uint256& hashTx, std::string& err_string, const CAmount& max_tx_fee, bool relay, bool wait_callback)
```
You could remove the `hashTx` parameter from this interface method. The only caller of the method supplies an unused uint256 and then throws it away.

jnewbery commented on src/interfaces/chain.cpp:
```diff
@@ -301,10 +296,10 @@ class ChainImpl : public Chain
-        CInv inv(MSG_TX, txid);
-        g_connman->ForEachNode([&inv](CNode* node) { node->PushInventory(inv); });
+        const TransactionError err = BroadcastTransaction(tx, txid, err_string, max_tx_fee, relay, false);
+        return TransactionError::OK == err || TransactionError::P2P_DISABLED == err;
```
This is exactly right, but I think it could do with a comment. Something like:

```
// Chain clients only care about failures to accept the tx to the mempool. Disregard non-mempool related failures
// Note: this will need to be updated if BroadcastTransaction() is updated to return other non-mempool failures
// that Chain clients do not need to know about.
```

Sjors replied:
It makes sense to filter some error messages, but for those that do make it through, at least the `err_string` - but probably also `TransactionError` - is useful for the client. So maybe it's better to change `TransactionError::OK` and erase the error message for errors that we don't want to propagate.

jnewbery reviewed (commented):
Looking really good. Just a few more nits!

ariard commented:
a00303c updated with @jnewbery last review comments, specially removing unused uint256 from BroadcastTransaction interface

ryanofsky commented:
Looks like this might also help a little bit with #14912, since right now it's adding another broadcastTransaction method (a7678372bd8b4ef4ce33a406559a0ea860ce1abf).

jnewbery commented:
Looks good. utACK e5b0fe904a9c6186b8273c61e93ee1538418b245

(I was only suggesting to remove the uint256 from the interface `broadcastTransaction()` method, not the `BroadcastTransaction()` function in node/transaction, but removing it there is also fine I think.)

Sjors reviewed (commented):
Concept ACK on adding this as an interface method (I haven't studied the rest). I helps me drop one commit from the hardware wallet support PR #14912.

It makes sense to filter some error messages in `BroadcastTransaction`, but for those that do make it through, at least the `err_string` - but probably also `TransactionError` - is useful for the client. So maybe it's better to change `TransactionError::OK` and erase the error message for errors that we don't want to propagate.

ariard commented:
@Sjors you mean erasing errors which seems to be used by no one right now (like TransactionError::MEMPOOL_ERROR) and just return a generic error for these cases ? Or returning errors from `broadcastTransactions` instead of a boolean and let caller to filter ?

Sjors commented:
@ariard I think it should return a `TransactionError`. You can filter it before returning, based on what you don't want callers to see. But not based on what's "used by no one right now".

jnewbery commented:
> You can filter it before returning, based on what you don't want callers to see. But not based on what's "used by no one right now".

I think it's fine to have the `broadcastTransaction` interface return what's most useful to our wallet right now. This interface is meant to be private, so having it customized to our needs is ok, and if future changes to the wallet require the interface method to be changed to return more information, that's also fine.

ariard commented:
Okay, well let it as of `0a1f4d9` with `broadcastTransaction` returning a boolean. But I checked failures, and seems like #15696 is causing functional test failures even locally..

Sjors commented:
Boolean is fine for my purposes as well.

ariard commented:
Rebased.

ajtowns commented on src/interfaces/chain.cpp:
```diff
@@ -295,6 +296,14 @@ class ChainImpl : public Chain
+        // Chain clients only care about failures to accept the tx to the mempool. Disregard non-mempool related failures.
+        // Note: this will need to be updated if BroadcastTransactions() is updated to return other non-mempool failures
+        // that Chain clients do not need to know about.
+        return TransactionError::OK == err || TransactionError::P2P_DISABLED == err;
```
Might be better to do:

```
switch(err) {
    case TransactionError::OK:
    case TransactionError::P2P_DISABLED:
        return true;
    case TransactionError::MISSING_INPUTS:
    case TransactionError::ALREADY_IN_CHAIN:
    case TransactionError::MEMPOOL_REJECTED:
    case TransactionError::MEMPOOL_ERROR:
    case TransactionError::INVALID_PSBT:
    case TransactionError::PSBT_MISMATCH:
    case TransactionError::SIGHASH_MISMATCH:
    case TransactionError::MAX_FEE_EXCEEDED:
        return false;
}
```

That way you get a warning if there's other values added that you haven't handled.

jnewbery replied:
I don't think it really matters too much. If new TransactionError values are added that the wallet needs to handle, this interface can be updated.

jnewbery committed:
Tidy up BroadcastTransaction()

jnewbery commented:
utACK 67c7db7a22e18cee5364652b9a1ab102c9fb1858

amitiuttarwar commented on src/node/transaction.cpp:
```diff
@@ -52,27 +53,27 @@ TransactionError BroadcastTransaction(const CTransactionRef tx, uint256& hashTx,
-    g_connman->ForEachNode([&inv](CNode* pnode) {
-        pnode->PushInventory(inv);
-    });
+    if (relay) {
```
could this use the relay txn method in [net_processing](https://github.com/bitcoin/bitcoin/blob/master/src/net_processing.cpp#L1308)?

ariard replied:
why not, see #16452

ariard replied:
why not, see #16452 because it can go on its own

jonatack commented on src/interfaces/chain.cpp:
```diff
@@ -290,10 +285,13 @@ class ChainImpl : public Chain
+        const TransactionError err = BroadcastTransaction(tx, err_string, max_tx_fee, relay, /*wait_callback*/ false);
+        // Chain clients only care about failures to accept the tx to the mempool. Disregard non-mempool related failures.
+        // Note: this will need to be updated if BroadcastTransactions() is updated to return other non-mempool failures
+        // that Chain clients do not need to know about.
```
Concept ACK / utACK. Reviewed, built, ran all unit/functional tests. Agree with @ajtowns concerning https://github.com/bitcoin/bitcoin/pull/15713#discussion_r306122680 to have a bit of a safety rail.

jonatack reviewed (commented):
utACK 67c7db7a22e18cee5364652b9a1ab102c9fb1858. Reviewed, built, ran all unit/functional tests. Agree with @ajtowns concerning https://github.com/bitcoin/bitcoin/pull/15713#discussion_r306122680 to have a bit of a safety rail.

amitiuttarwar commented:
utACK 67c7db7a22e18cee5364652b9a1ab102c9fb1858. Carefully read the code, built & ran all tests.

ariard commented:
New tip at c1351e2 with @jnewbery BroadcastTransaction cleaning commit

jnewbery commented:
ACK c1351e28fd2e38a928b322456038198aa0e847bd

(but I don't think it's necessary to take that commit and invalidate @amitiuttarwar and @jonatack reviews)

jonatack commented:
Partial ACK 976d1e4858df00c031ef2c07790862948dbc04cf, all tests pass, local git range-diff shows no change in previously-ACKed commits.

New commit makes doc-only changes in `src/node/transaction.h` and changes BroadcastTransaction logic in `src/node/transaction.cpp`. IIUC the logic changes are:

before
```
if txn not in mempool and coin is spent
  submit txn to local node and sync with wallets
else if coin is not spent
  return "already in chain" txn error
else txn is in mempool and coin is spent
  ensure we don't block if re-sending a transaction already in mempool
end
```

now
```
exit loop early if coin is not spent and return "already in chain" txn error

if txn not in mempool
  submit txn to local node and sync with wallets
end
```

jnewbery commented:
ACK 976d1e485

promag commented on src/rpc/rawtransaction.cpp:
```diff
@@ -807,14 +807,13 @@ static UniValue sendrawtransaction(const JSONRPCRequest& request)
-    uint256 txid;
     std::string err_string;
-    const TransactionError err = BroadcastTransaction(tx, txid, err_string, max_raw_tx_fee);
+    const TransactionError err = BroadcastTransaction(tx, err_string, max_raw_tx_fee, true, true);
```
87cca293d5b127e43e35d866167893454658eacd

nit, could add a comment on the last 2 args.

promag commented on src/node/transaction.cpp:
```diff
@@ -52,27 +53,27 @@ TransactionError BroadcastTransaction(const CTransactionRef tx, uint256& hashTx,
+            return TransactionError::P2P_DISABLED;
+        }
+        CInv inv(MSG_TX, hashTx);
+        g_connman->ForEachNode([&inv](CNode* pnode) {
```
87cca293d5b127e43e35d866167893454658eacd

just noting that this must be changed in #16452 to `RelayTransaction`.

promag commented on src/node/transaction.h:
```diff
@@ -14,11 +14,13 @@
- * @param[in]  highfee Reject txs with fees higher than this (if 0, accept any fee)
+ * @param[in]  max_tx_fee reject txs with fees higher than this (if 0, accept any fee)
+ * @param[in]  relay flag if both mempool insertion and p2p relay are requested
+ * @param[in]  wait_callback, wait until callbacks have been processed to avoid stale result due to a sequentially RPC.
```
87cca293d5b127e43e35d866167893454658eacd

I think a better description/use-case is needed here.

jnewbery replied:
I think that the comments here: https://github.com/bitcoin/bitcoin/pull/15713/files#diff-3f841b5d557d579ef74c43ff582b16ebR16 and here: https://github.com/bitcoin/bitcoin/pull/15713/files#diff-bf669d26afe05f397f3f01bad0c1ff0eR52 are sufficient.

Sjors commented on src/node/transaction.cpp:
```diff
@@ -21,15 +21,15 @@ TransactionError BroadcastTransaction(const CTransactionRef tx, std::string& err
-    bool fHaveChain = false;
-    for (size_t o = 0; !fHaveChain && o < tx->vout.size(); o++) {
+    for (size_t o = 0; o < tx->vout.size(); o++) {
         const Coin& existingCoin = view.AccessCoin(COutPoint(hashTx, o));
```
Nit: add comment explaining `IsSpent()` doesn't mean the coin is spent; it means the output doesn't exist. So if the output does exist, then this transaction exists in the chain, i.e. it's already confirmed.

jnewbery replied:
Agree that a comment here would be helpful in a follow-up commit.

ariard replied:
Yes agree better documentation and refactoring of BroadcastTransaction can be pushed further in latter work.

Sjors reviewed (approved):
Code review ACK 976d1e4.

As to why dropping `if (GetDepthInMainChain(locked_chain) != 0) return false;` is safe, IIUC because `BroadcastTransaction` loops through the transaction outputs and checks if they exist at the chain tip. By the way, the use of `AccessCoin` and `IsSpent` (actually meaning: "does not exist") in `BroadcastTransaction` seems a needlessly confusing optimization.

mzumsande commented:
I'm not sure if `broadcastTransaction` should suppress `TransactionError::P2P_DISABLED`. 
From the old discussion with @jnewbery above I gather that this is the case because the old `relayTransaction()` didn't have a return code, and so that we can update `fInMempool` conveniently.

However, the old `CWalletTx::RelayWalletTransaction()`  had an explicite check for P2P (`if (!pwallet->chain().p2pEnabled()) return false;`) so if the node was not connected to the p2p network it would not call `relayTransaction()` and return false.  Its replacement function SubmitMemoryPoolAndRelay doesn't have this check and will return true in this case, so this looks like a change in behavior ( minor because its return value is only used for log messages as far as I can see).

jnewbery reviewed (commented):
> Its replacement function SubmitMemoryPoolAndRelay doesn't have this check and will return true in this case, so this looks like a change in behavior ( minor because its return value is only used for log messages as far as I can see).

Great catch @mzumsande! I agree that this is a slight change in behaviour, but that it's only used for logging in `ResendWalletTransactions()` (The return value in `CommitTransaction()` should be the same as before).

ariard commented:
Thanks for review @mzumsande, I agree that's a light change in behavior even it was only logs. IMO `p2pEnabled` is just there when wallet is init before `g_connman`, and a quick look at `init.cpp` I'm not even sure that's something possible. In future works, I'm thinking to remove `p2pEnabled` once we have the wallet registering its current height with Chain interface at `handleNotifications`. Wallet will only submit tx to the mempool and shouldn't be aware of P2P state.

mzumsande reviewed (commented):
Thanks for the answer. In light of the intended future direction it makes sense not to implement a more fine-grained error handling with separate errors for mempool and relay.
ACK 976d1e4858df00c031ef2c07790862948dbc04cf (reviewed the code and ran tests).

maflcko reviewed (commented):
ACK 976d1e4858df00c031ef2c07790862948dbc04cf, some questions

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 976d1e4858df00c031ef2c07790862948dbc04cf, some questions
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiQegv/UAc28xrmuUqCZFUkvZtoLuw2Bq7qPa5zsh4Qr63OKNgh+BlzIh02YB3E
l+8XvwKaPIhEk+flUc5KrBHpCe5kbRe0sY3X1FFVCI5ENTSKmm2+9U81ISgEAaZR
nrZQL3kiwjSXN65k4lH3QfDgLJn5+LDZbW++Q/0NMFczMOPufEWZ168YXaZVnjpn
Ts5cqHG4wfCdpbWVq8fFLxIYSKQEbu55+nheGG2dbDKu+qVSxt1J2+1NnAsVWHmh
PTaUg9tmZqtpZuagHO7wGZj2Iwk7XZFjwi2htbKqoHjrOlov21FFZkcslAgc6Go4
tvz0S0Bo5Y2mGRJ4wrioTfoDs0WFci2zBFdE1EQ11+hSuDGTcny4CKcnSFmlRWlc
BD/34jM+55HM0okGs8lk6USeDsC7pn9Lzka6VOgM6D/N+O9nekimQvACdtY4/FHl
5fFiZp4Jc6hiYMIqyJsy+ooFufQTX6pVvGBrk7OVBE9GvCWFLAbv+arfO+1z4UUU
uk0WUV1R
=wLf3
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `27c1d332b653a5bb053c1a397b8772ee22e02d0e8091c6dfbe094774aaab128f  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010827c1d332b653a5bb053c1a397b8772ee22e02d0e8091c6dfbe094774aaab128ff010d1fd04d7df2e13db8d24cf852915534c08fff01044023faf81875a0a98c864deaf77f4a108f1045d40698df008accf899bd118fa740083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010b34035f3d96978562da20080ea7c975f08f1045d40698df008307826ac98988fa20083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010bd81ac0e834c17d7b91e87433667751208f1045d40698df0086631396e8e93e6c10083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010ff7ace59e3ccbce8f8d11333b091a4de08f1045d40698df0089bb14ee9dbceffca0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

ariard commented on src/node/transaction.h:
```diff
@@ -11,14 +11,21 @@
+ * Mempool submission can be synchronous (will await mempool entry notification
+ * over the CValidationInterface) or asynchronous (will submit and not wait for
+ * notification), depending on the value of wait_callback. wait_callback MUST
+ * NOT be set while cs_main, cs_mempool or cs_wallet are held to avoid
```
Yes, we should do that in `sendrawtransaction`, which is out of scope for this PR (but glad to address it in another PR if you think that's relevant)

MarcoFalke replied:
Ok, lets do this in a follow-up

ariard replied:
@MarcoFalke should I include headers for mempool and wallet to `AssertLockNotHeld` these locks too ?

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -2353,7 +2352,8 @@ void CWallet::ResendWalletTransactions()
             if (wtx.nTimeReceived > m_best_block_time - 5 * 60) continue;
-            if (wtx.RelayWalletTransaction(*locked_chain)) ++relayed_tx_count;
+            std::string unused_err_string;
+            if (wtx.SubmitMemoryPoolAndRelay(unused_err_string, true, *locked_chain)) ++relayed_tx_count;
```
I guess we should recall it `submitted_count`  ? Wallet is just submitting to mempool and that's mempool job to pass it to P2P. We don't dissociate errors here anyways

MarcoFalke replied:
I don't care. It is only a log print, so can be done later. Though, I wanted to point out the change in behavior, since you tagged this with "refactor".

ariard replied:
Yes sorry for that, I modified the PR a bit since opening, so it's no more a refactor I guess..

MarcoFalke replied:
Lets do this in a follow-up, if important enough

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -2125,20 +2125,16 @@ void CWallet::ReacceptWalletTransactions(interfaces::Chain::Lock& locked_chain)
     // Can't relay if wallet is not broadcasting
     if (!pwallet->GetBroadcastTransactions()) return false;
-    // Don't relay coinbase transactions outside blocks
-    if (IsCoinBase()) return false;
```
You mean compare to the one at beginning of ATMP ?

MarcoFalke replied:
`BroadcastTransaction` does a convenience check that is only used in the sendrawtransaction (to notify a user of a transaction that is already confirmed). This is done by looking up if at least one of the outputs is in the utxo set. For the purpose of the wallet, this check is completely useless. Previously we wouldn't call it and now we do (even for coinbase txs), where it can be trivially avoided.

ariard replied:
Ah yes, should the wallet try to check as much as confirmation rules it can, like the coinbase one ? It's a layer violation, but that's trivial as you said.

MarcoFalke replied:
Why is it a layer violation of the wallet to call a [pure method](https://en.wikipedia.org/wiki/Pure_function) on one of it's txs?

ariard replied:
Really IMO, it's a layer violation because the wallet is doing the job of the node to check consensus rules instead of relying on what the node says.

ariard replied:
Do you want I update to get back the check ? Wallet are not going to broadcast thousands of coinbase transactions to this be real performance issue.

jnewbery replied:
I agree with Marco that this isn't a layer violation. `IsCoinBase()` is a pure utility function and can be called anywhere.

But I also agree with ariard that this isn't a big deal. Calling `IsCoinBase()` here is an optimization that affects a tiny number of transactions for a tiny number of users.

promag replied:
Reverted in #16557.

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -2140,17 +2139,17 @@ bool CWalletTx::RelayWalletTransaction(interfaces::Chain::Lock& locked_chain)
-    return true;
+    // Submit transaction to mempool for relay
+    pwallet->WalletLogPrintf("Submitting wtx %s to mempool for relay\n", GetHash().ToString());
+    // We must set fInMempool here - while it will be re-set to true by the
```
Noted if I have to rebase, or if I can throw it in one of my other PR touching this code

MarcoFalke replied:
Lets do this in a follow up

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -2125,32 +2124,28 @@ void CWallet::ReacceptWalletTransactions(interfaces::Chain::Lock& locked_chain)
     // Don't relay abandoned transactions
     if (isAbandoned()) return false;
-    // Don't relay conflicted or already confirmed transactions
-    if (GetDepthInMainChain(locked_chain) != 0) return false;
```
You mean relying on `AccessCoin` in `BroadcastTransaction`. Isn't `CCoinsViewCache` a in-memory cache? Is the wallet broadcasting that much transactions so it can be a performance penalty ? `GetDepthInMainChain` is also currently locking the chain via `getBlockHeight`

Sjors replied:
`CCoinsViewCache` hits disk if the in memory cache doesn't contain the coin we're looking for. I doubt the performance matters for the wallet, unlike e.g. peer to peer mempool relay.

MarcoFalke commented:
See a patch (on top of master and this pull) for what I was thinking:

[0001-Check-utxo-db-early-only-when-requested-in-Broadcast.patch.txt](https://github.com/bitcoin/bitcoin/files/3451955/0001-Check-utxo-db-early-only-when-requested-in-Broadcast.patch.txt)

MarcoFalke commented on src/interfaces/chain.cpp:
```diff
@@ -290,10 +285,13 @@ class ChainImpl : public Chain
     {
-        CInv inv(MSG_TX, txid);
-        g_connman->ForEachNode([&inv](CNode* node) { node->PushInventory(inv); });
+        const TransactionError err = BroadcastTransaction(tx, err_string, max_tx_fee, relay, /*wait_callback*/ false, /*check_chain*/ false);
```
Sorry, I think I got this wrong in my patch

```suggestion
        const TransactionError err = BroadcastTransaction(tx, err_string, max_tx_fee, relay, /*check_chain*/ false, /*wait_callback*/ false);
```

ariard commented:
Okay pushed with your patch amended to invert comment on 4441ba6c. 

More I looking at `BroadcastTransaction`, more I think it needs a PR of its own, maybe I can take your commit here?

jnewbery commented:
I don't like Marco's suggested change. It's leaking implementation detail to the caller and complicating the function interface. I also don't understand the problem it's supposed to be solving - is it just a performance optimization so that we don't do the UTXO check for wallet-submitted transactions?

This PR had a bunch of ACKs without that change. Does it need to be included before this PR can be merged?

MarcoFalke commented:
> This PR had a bunch of ACKs without that change. Does it need to be included before this PR can be merged?

No, but I'd like to hear other's opinion on it.

Also, you have just requested the conflicting pull to be merged, so this needs re-ACK anyway:

* https://github.com/bitcoin/bitcoin/pull/16452#issuecomment-516871453

ariard commented:
Rebased at 64b16d2 with https://github.com/bitcoin/bitcoin/pull/15713#discussion_r308697103, https://github.com/bitcoin/bitcoin/pull/15713#discussion_r308796603, https://github.com/bitcoin/bitcoin/pull/15713#discussion_r308798753, https://github.com/bitcoin/bitcoin/pull/15713#discussion_r307831700, https://github.com/bitcoin/bitcoin/pull/15713#discussion_r307813745 addressed. Shouldn't have functional changes since 976d1e4.

Please review again, and let give your opinion on Marco patch : https://github.com/bitcoin/bitcoin/pull/15713#issuecomment-516842095.

jnewbery commented:
You seem to have lost the `if (relay)` conditional in `BroadcastTransaction()`. I think that's a rebase error.

You've also lost the `g_connman` error checking block:

```
        if (!g_connman) {
            return TransactionError::P2P_DISABLED;
        }
```

Alternatively, you could assert that `g_connman` is non-null at the top of this function. It's only called by RPCs and the wallet scheduler, both of which are initialized after `g_connman` is assigned and stopped before `g_connman` is reset.

ariard commented:
Oooops, thanks to pointing me the error, remove the check to be replaced by an assertion as it's less checks in BroadcastTransaction at fb62f12.

Note : currently P2P_DISABLED error is still there even if it's not used. It can be clean in a later commit if we agree that disabling P2P is not a thing.

MarcoFalke commented:
re-ACK fb62f128bbfd8c6cd72ea8e23331a4bae23883ab

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK fb62f128bbfd8c6cd72ea8e23331a4bae23883ab
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgBCAv/bnyyRCCqUH8PSxg2lces1K5Z5M/Nb+eT5O2JQhQb1EfY+NckIcAqPmZz
ffhn6qmibcfEUWVnhBdGD/8xZOdUM37IPs1xCVFt7kYdTIDkLGKKZqg9z9R14CZd
aivpQia09e6zURD6a/d7qB8isfRCTNe3a29dcZv43oW2+WIStQGRBGswymI17k7C
fYPsfWkAIru83VnoPCeziwXR85mRTXmKfVSQbgM/Qned2NDGGqZRZ6YH8KdtyPpP
e3p0EAYMB+6aXum3Ky+u5gu9yRbx5pjBebBzQKbHHFHMLXxIBKAHml6qK7zqiO87
Sg5fSDEQ/NC2DFfarRVWNT36UAhUqxpKwQptpWzMvgLXc2HYOR6MQrKM+eg/x84M
YN5b9zlKXGySUWVZWqk+mL9ELUPDxZu6zRtzEK1TI8WXQi+bqqQrjV6rp8ZqGj+s
M11E+efGRFL1qYOMEs0lHHcdqODkUoOJP1csVAGTDvFNJSDvbHMuqGUuObCHU8V1
uqYjCCQV
=XQET
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `9cd92754a615d69deebd843e2a0522965f52b1cf18a0916f147bb00a450d5790  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401089cd92754a615d69deebd843e2a0522965f52b1cf18a0916f147bb00a450d5790f0109414c84155f97676e56ab1908863926808fff0107fd1fefd33857ac615713e5ded92b7a508f1045d432706f0089eea0e4fbf0bbe6e0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0109168bac465e85f73ba3c118b4dcef2dd08f1045d432705f008d0be8a4720a55d4c0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010ad8d8db57fd4036e0e615c9f7d967bda08f1045d432706f0089036534856f553e30083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010eaf159045992815a83a18f4aecce73b008f1045d432706f008f4e5d172ca8b65ab0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

Sjors commented on src/node/transaction.cpp:
```diff
@@ -53,24 +55,21 @@ TransactionError BroadcastTransaction(const CTransactionRef tx, uint256& hashTx,
 
-    promise.get_future().wait();
-
-    if (!g_connman) {
```
This check was added by @theuni in 53347f0cb99e514815e44a56439a4a10012238f8. Why wasn't this an assert in the first place?

Nit: this change could use its own commit.

Sjors commented on src/node/transaction.cpp:
```diff
@@ -14,26 +14,30 @@
+    for (size_t o = 0; o < tx->vout.size(); o++) {
         const Coin& existingCoin = view.AccessCoin(COutPoint(hashTx, o));
-        fHaveChain = !existingCoin.IsSpent();
+        // IsSpent doesnt mean the coin is spent, it means the output doesnt' exist.
```
Nit: misspelled `doesn't` twice.

Sjors reviewed (approved):
re-ACK fb62f128bbfd8c6cd72ea8e23331a4bae23883ab

Let's not bother with the `check_chain` patch 4441ba6cd940cf70b59182373cefeb28c22da4be (it's not in the current version of this PR).

MarcoFalke commented:
This (seemingly simple) change has already 132 comments. I am going to merge it now, because I believe it is a step in the right direction. We can fix up typos and other stuff later, if needed.

jnewbery commented on src/interfaces/chain.cpp:
```diff
@@ -291,9 +286,13 @@ class ChainImpl : public Chain
     {
-        RelayTransaction(txid, *g_connman);
+        const TransactionError err = BroadcastTransaction(tx, err_string, max_tx_fee, relay, /*wait_callback*/ false);
+        // Chain clients only care about failures to accept the tx to the mempool. Disregard non-mempool related failures.
```
I think this line can be removed now. The only failures that `BroadcastTransaction()` can return are mempool-related failures.

jnewbery commented on src/node/transaction.cpp:
```diff
@@ -14,26 +14,30 @@
-TransactionError BroadcastTransaction(const CTransactionRef tx, uint256& hashTx, std::string& err_string, const CAmount& highfee)
+TransactionError BroadcastTransaction(const CTransactionRef tx, std::string& err_string, const CAmount& max_tx_fee, bool relay, bool wait_callback)
 {
+    assert(g_connman);
```
nit: I'd prefer to see a comment next to this assert to explain why `g_connman` must always be assigned here.

MarcoFalke replied:
Should be part of #16503, maybe

jnewbery reviewed (commented):
post-merge ACK fb62f128bbfd8c6cd72ea8e23331a4bae23883ab.

A couple of nits inline that could be cleaned up in a subsequent PR.

jnewbery commented:
I was wrong about this: https://github.com/bitcoin/bitcoin/pull/15713#discussion_r309244186. Apparently removing that check causes log spam, particularly for test nodes.
